### DISCUSSION TITLE
How can I solve this problem, "undefined reference to `CppAD::local::temp_file()`" ?

### DISCUSSION THREAD

**Discussion by MIYATANIKoki** [coin-or/CppAD#182] [Q&A] [answered]
Hello, I have a question regarding the installation of cppad.
I'd appreciate if you answer the questions below.

I am trying to create an MPC using cppad and ipopt in ROS.
The OS is ubuntu 20.04.

I have installed CppAD using the following steps.

1. download CppAD
    in ${DIR}, the following command was executed.
    $ git clone https://github.com/coin-or/CppAD.git cppad.git

2. download optional packages
    First, I installed blas and lapack.

    And in ${DIR}/cppad.git, the following commands were executed.
    $ mkdir build
    $ mkdir local
    $ bin/get_adolc.sh
    $ bin/get_eigen.sh
    $ bin/get_ipopt.sh

3. cmake & install
    First, ${DIR}/cppad.git/build/prefix/lib/pkgconfig and 
    ${DIR}/cppad.git/build/prefix/lib64/pkgconfig were added to PKG_CONFIG_PATH.

    And in ${DIR}/cppad.git/build, 
    $ cmake                                                                      \
          -D CMAKE_VERBOSE_MAKEFILE=true                                         \
          -G "Unix Makefiles"                                                    \
          -D cppad_prefix=${DIR}/cppad.git/local                                 \
          -D cmake_install_includedirs=include                                   \
          -D cmake_install_libdirs=lib                                           \
          -D include_adolc=true                                                  \
          -D include_eigen=true                                                  \
          -D include_ipopt=true                                                  \
          -D cppad_testvector=eigen                                              \
          ..
    
    $ make check_cppad_ipopt
    All tests have passed.
    
    $ make install
    includes and libs of cppad were installed in ${DIR}/cppad.git/local.

4. Install includes and libs related to CppAD in usr/local
    Copied cppad from ${DIR}/cppad.git/local to /usr/local.
    Also copied adolc and coin-or from ${DIR}/cppad.git/build/prefix to /usr/local.

MPC was written with reference to ipopt_solve_get_started.cpp.
And my CmakeList.txt is as follows.

cmake_minimum_required(VERSION 3.0.2)
project(nlp_solver)

add_definitions(-std=c++11 -O3)

set(CXX_FLAGS "-Wall")
set(CMAKE_CXX_FLAGS, "${CXX_FLAGS}")

find_package(catkin REQUIRED COMPONENTS
  roscpp
  rospy
  std_msgs
  custom_msgs
)

find_package(Eigen3 3.3 REQUIRED
)

catkin_package()

include_directories(
  ${catkin_INCLUDE_DIRS}
  include
  /usr/include/eigen3
  /usr/local/include
)

link_directories(
  /usr/local/lib
  /usr/local/lib64
)

add_executable(nlp_node
  src/main.cpp
  src/nlp_solver.cpp
)

target_link_libraries(nlp_node
  ${catkin_LIBRARIES}
  Eigen3::Eigen
  ipopt
)

I did a catkin build in my ROS workspace and detected following errors.

/usr/bin/ld: CMakeFiles/nlp_node.dir/src/nlp_solver.cpp.o: in function `void CppAD::put_check_for_nan<double>(CppAD::vector<double> const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&)':
nlp_solver.cpp:(.text._ZN5CppAD17put_check_for_nanIdEEvRKNS_6vectorIT_EERNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEE[_ZN5CppAD17put_check_for_nanIdEEvRKNS_6vectorIT_EERNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEE]+0x4f): undefined reference to `CppAD::local::temp_file[abi:cxx11]()'
collect2: error: ld returned 1 exit status

How can I solve this problem, "undefined reference to `CppAD::local::temp_file()`" ?
I would be happy to answer any questions you may have.

**Comment by bradbell** [comment, marked as answer]
temp_file should be in cppad_lib; see
https://github.com/coin-or/CppAD/blob/master/cppad_lib/temp_file.cpp

**Comment by MIYATANIKoki** [comment]
Sorry, it's my mistake.
I added cppad_lib in target_link_libraries, then catkin build succeeded.

Thank you for your quick response.
